main.go

```diff
@@ -0,0 +1,66 @@
+package main
+
+import (
+	"log"
+	"net"
+	"net/http"
+
+	"github.com/gorilla/websocket"
+)
+
+const (
+	readBufferSize  = 1024
+	writeBufferSize = 1024
+	maxMessageSize  = 1024
+)
+
+func handler(clients map[*Client]bool, gameInChannel chan InFrame, w http.ResponseWriter, r *http.Request) {
+	upgrader := websocket.Upgrader{
+		ReadBufferSize:  readBufferSize,
+		WriteBufferSize: writeBufferSize,
+		CheckOrigin: func(r *http.Request) bool {
+			return true
+		}}
+
+	conn, err := upgrader.Upgrade(w, r, nil)
+	if err != nil {
+		log.Println("upgrade: ", err)
+	}
+
+	client := NewClient(conn)
+	clients[client] = true
+
+	go client.tx()
+	go client.rx(gameInChannel)
+}
+
+func main() {
+	log.SetFlags(log.LstdFlags | log.Lshortfile)
+
+	// Set of all connected clients
+	clients := make(map[*Client]bool)
+
+	// Create the game inbound channel
+	gameInChannel := make(chan InFrame)
+	gameOutChannel := make(chan OutFrame)
+
+	// Connect to the game server
+	// TODO: This is temporary; in reality this should be some protobuf over TCP
+	gameConn, _ := net.Pipe()
+
+	// Create the game dispatchers
+	go game_tx(gameOutChannel, gameConn)
+	go game_rx(gameInChannel, gameConn)
+
+	// Set up the listener
+	http.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
+		handler(clients, gameInChannel, w, r)
+	})
+
+	// Start listening
+	// TODO: Read the listen host from command-line or config or something
+	err := http.ListenAndServe("localhost:4000", nil)
+	if err != nil {
+		log.Fatal(err)
+	}
+}
```

Comment:
Yes, this is idiomatic Go implementation but not optimized. Consider 1 mln users. The client sends a message/packet with the relevant events or in other words subscribes for events. Then (not taking into account technical messages such as ping/pong), the client may send nothing else for the whole connection lifetime. The connection lifetime may last from several seconds to several days. So for the most time rx() and tx() are waiting for the handling of data for receiving or sending. Along with them waiting are the I/O buffers of 4 KB each. And go routines  with 2KB stacks each.

For scalability you should use netpoll mode and get rid of goroutines by creating go routines pool.  Something like this. https://godoc.org/github.com/mailru/easygo/netpoll 
Another point is HTTP-upgrade mode. Each sync requires copy. You can use something like github.com/gobwas/ws

Conclusion:

For playing around this is fine. For serious step ahead, need to cosnsider the following points.
rx() goroutine with a buffer inside is expensive -> netpoll (epoll, kqueue); reuse the buffers.
tx() goroutine with a buffer inside is expensive -> start the goroutine when necessary; reuse the buffers.
With burst of connections, netpoll won’t work -> reuse the goroutines with the limit on their number.
net/http is not the fastest way to handle Upgrade to WebSocket -> use the zero-copy upgrade on bare TCP connection.

Reply:
https://medium.freecodecamp.org/million-websockets-and-go-cc58418460bb makes good points; actually I was planning to do a pool of goroutines initially, but then based on some of our conversations I felt like we had concluded that one goroutine per connection was a fine approach. The article does make it clear though that the pool-based approach with epoll is more performant. I'll see what if I can do something interesting in that direction here.